sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -1164,4 +1167,27 @@ object DateTimeUtils {
     threadLocalTimestampFormat.remove()
     threadLocalDateFormat.remove()
   }
+
+  class MicrosCalendar(tz: TimeZone) extends GregorianCalendar(tz, Locale.US) {
+    def getMicros(digitsInFraction: Int): SQLTimestamp = {
+      val d = fields(Calendar.MILLISECOND) * MICROS_PER_SECOND
+      d / Decimal.POW_10(digitsInFraction)
```

Comment:
This is really mysterious. How does it work?

Reply:
AFAIK `fields(Calendar.MILLISECOND)` is simply the result of parsing the digits after dot to int. For example:
- `.1234` returns 1234
- `.100` returns 100
- `.01` returns 1

The last case is tricky as we don't know what the original string is. e.g. both `.01` and `.1` returns 1. How can we correct the value in this case?

Reply:
> The last case is tricky as we don't know what the original string is ...

Right, we don't know but we know how many digits must be. We count number of 'S' in the pattern: https://github.com/apache/spark/pull/26507/files#diff-da60f07e1826788aaeb07f295fae4b8aR1191

For the pattern `SS`, the algorithm work correctly only when the input string has at least 2 digits in the seconds fraction. If you remove trailing zeros, the result is incorrect because we cannot distinguish `.01` and `.1`. So, inputs for the pattern `SS` must be `.01`, `.10`, `.00`, `.99` and so on.

Existing round trip tests in JSON and CSV datasource are passed successfully because `SimpleDateFormat/FastDateFormat` produce the same number of digits as specified in the pattern without removing trailing zeros.

Actually, the same rule for number of digits is applicable for parsing milliseconds **without my changes** , see my comment: https://github.com/apache/spark/pull/26507#discussion_r345982915 . Besides of supporting microseconds, this PR also supports patterns `S`, `SS`, `SSSS`, `SSSSS`. 2.4 works correctly only for `SSS` and number of digits in the input string equals to 3, nothing more.

Reply:
> This is really mysterious. How does it work?

The idea is similar to what we have for parsing interval's seconds with the fractional part. Let's say the pattern is `SSSS` and the input is `1234`:
1. Add 6 zeros at the end for the worse case: `1234` -> `1234000000`
2. Take the first 6 digits  `1234000000` -> `123400`. The rest is `0000` with the length is the number of digits in the pattern. Actually `123400` = `1234000000` / 10 ^ 4

Reply:
> So, inputs for the pattern SS must be .01, .10, .00, .99 and so on.

What if it's not satisfied? Will we fail or return wrong anwser?

To be clear I'm fine with it. It's better than before. At least we give users a way to get the correct result, by setting the format string correctly.

Reply:
If an user sets the pattern incorrectly, the result will be incorrect too. Before the changes, we can observe the same behavior, see https://github.com/apache/spark/pull/26507#discussion_r345982915 . The difference between old and new behavior is we allow to set a correct fraction patterns `S`, `SS`, `SSS`, `SSSS`, `SSSSS`, and `SSSSSS`.